### ISSUE TITLE
Socket not receiving response from the server ...

### ISSUE THREAD

**Issue by Arunbalaji001** [socketio/socket.io-client-swift#985]
Hi 
  I am using your library in objective c, the following code works finely on emitting a socket to the server, but it doesn't receive the sockets emitted from the server.The android client is working fine with the same format of coding so I don't think the error is with the naming of sockets. I am having trouble with "socket on" function only.If possible kindly share a solution to this.

**Objective C client**

    NSURL* url = [[NSURL alloc] initWithString:@"http://192.168.0.109:3000"];
       [socket on:@"connection" callback:^(NSArray* data, SocketAckEmitter* ack) {
        NSLog(@"socket connected");
        // [socket emit:@"setUsername" with:typea];
        //   [socket emit:@"setUsername" with:@[@"test"]];
        // [manager emitAll:@"setUsername" withItems:@[@"tulasi"]];
        [manager emitAll:@"join_user" withItems:[SharedData sharedConstants].commonuserid];
        NSLog(@"socket connected:%@",[SharedData sharedConstants].commonuserid);
    }];
    
        //    NSArray * idarray =[NSString stringWithFormat:@"@%@",[SharedData             sharedConstants].commonuserid];
          //    NSLog(@"id array is :%@",idarray);
         //
        //
      //[manager emitAll:@"join_user" withItems:@[@"commonuserid"]];
    
    
    
    manager= [[SocketManager alloc] initWithSocketURL:url config:@{@"log": @YES, @"compress":            @YES}];  
    
   
    
    socket = manager.defaultSocket;
    
    
    [self->socket on:@"my_message" callback:^(NSArray* data, SocketAckEmitter* ack) {
        NSLog(@"join=>%@", data.description);
        messagestr =[[data valueForKey:@"message"]objectAtIndex:0];
       // NSString * names =[[data valueForKey:@"userfrom"]objectAtIndex:0];

        NSLog(@"receive response is:%@",messagestr);

        [chatarray addObject:messagestr];

        [typearr addObject:@"b"];
        NSLog(@"left array is:%@",chatarray);
        NSLog(@"type array is:%@",typearr);
   
   
    
    }];
    
    
    
    [socket on:@"my_message" callback:^(NSArray * _Nonnull arrayData, SocketAckEmitter * _Nonnull ack) {
        
        //NSLog(@"test:%@", [arrayData description]);
        //         [self updateTableView];
         //double cur = [[arrayData objectAtIndex:0] floatValue];
        messagestr =[[arrayData valueForKey:@"message"]objectAtIndex:0];
      //  NSString * names =[[arrayData valueForKey:@"userfrom"]objectAtIndex:0];
        
        NSLog(@"receive response message is:%@",messagestr);
       // NSLog(@"receive response name is :%@",names);
        
        
        [chatarray addObject:messagestr];
        
        [typearr addObject:@"b"];
        NSLog(@"left array is:%@",chatarray);
        NSLog(@"type array is:%@",typearr);
        
        //message
        
        
        //typing
        
        
        
        
        
        [self updateTableView];
        
    }];
    NSLog(@"didReceiveEvent()");
    
    
    
    [self->socket connect];
    
}

**Comment by nuclearace** [comment]
Is the manager a property? Can you post some logs?

**Comment by Arunbalaji001** [comment]
SocketManager* manager;
    SocketIOClient* socket;

"SocketManager" is declared as the manager and the "SocketIoClient" is declared as socket.

In the console, it is showing the are messages sent but not showing anything about the socket received.

**Logs :**

018-03-26 16:49:18.942306 datelah[2324:754465] LOG SocketIOClient{/}: Joining namespace /
2018-03-26 16:49:18.942966 datelah[2324:754465] LOG SocketManager: Tried connecting socket when engine isn't open. Connecting
2018-03-26 16:49:18.943399 datelah[2324:754465] LOG SocketManager: Adding engine
2018-03-26 16:49:25.386674 datelah[2324:754883] LOG SocketEngine: Starting engine. Server: http://192.168.0.109:9005
2018-03-26 16:49:25.386935 datelah[2324:754883] LOG SocketEngine: Handshaking
2018-03-26 16:49:25.388457 datelah[2324:754883] LOG SocketEnginePolling: Doing polling GET http://192.168.0.109:9005/socket.io/?transport=polling&b64=1
2018-03-26 16:49:25.677721 datelah[2324:754891] LOG SocketEnginePolling: Got polling response
2018-03-26 16:49:25.679886 datelah[2324:754891] LOG SocketEnginePolling: Got poll message: 97:0{"sid":"cMDOxgYsL68hrm5AAAAZ","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}
2018-03-26 16:49:25.691134 datelah[2324:754891] LOG SocketEngine: Got message: 0{"sid":"cMDOxgYsL68hrm5AAAAZ","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}
2018-03-26 16:49:25.722136 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: ping with data: []
2018-03-26 16:49:25.722654 datelah[2324:754465] LOG SocketManager: Engine opened Connect
2018-03-26 16:49:25.722934 datelah[2324:754891] LOG SocketEnginePolling: Doing polling GET http://192.168.0.109:9005/socket.io/?transport=polling&b64=1&sid=cMDOxgYsL68hrm5AAAAZ
2018-03-26 16:49:25.722977 datelah[2324:754465] LOG SocketIOClient{/}: Socket connected
2018-03-26 16:49:25.723278 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: statusChange with data: [connected]
2018-03-26 16:49:25.723748 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: connect with data: ["/"]
2018-03-26 16:49:25.724306 datelah[2324:754891] LOG SocketEngine: Writing poll:  has data: false
2018-03-26 16:49:25.727234 datelah[2324:754891] LOG SocketEnginePolling: Sending poll:  as type: 2
2018-03-26 16:49:25.732347 datelah[2324:754891] LOG SocketEnginePolling: Created POST string: 1:2
2018-03-26 16:49:25.733603 datelah[2324:754891] LOG SocketEnginePolling: POSTing
2018-03-26 16:49:25.733990 datelah[2324:754891] LOG SocketEnginePolling: Doing polling POST http://219.75.20.92:9005/socket.io/?transport=polling&b64=1&sid=cMDOxgYsL68hrm5AAAAZ
2018-03-26 16:49:25.778918 datelah[2324:754893] LOG SocketEnginePolling: Got polling response
2018-03-26 16:49:25.779374 datelah[2324:754893] LOG SocketEnginePolling: Got poll message: 2:40
2018-03-26 16:49:25.779980 datelah[2324:754893] LOG SocketEngine: Got message: 40
2018-03-26 16:49:25.780463 datelah[2324:754465] LOG SocketParser: Parsing 0
2018-03-26 16:49:25.782125 datelah[2324:754465] LOG SocketParser: Decoded packet as: SocketPacket {type: 0; data: []; id: -1; placeholders: 0; nsp: /}
2018-03-26 16:49:25.782280 datelah[2324:754893] LOG SocketEnginePolling: Doing polling GET http://192.168.0.109:9005/socket.io/?transport=polling&b64=1&sid=cMDOxgYsL68hrm5AAAAZ
2018-03-26 16:49:25.830647 datelah[2324:754892] LOG SocketEngineWebSocket: Sending ws: probe as type: 2
2018-03-26 16:49:25.840207 datelah[2324:754893] LOG SocketEnginePolling: Got polling response
2018-03-26 16:49:25.840511 datelah[2324:754893] LOG SocketEnginePolling: Got poll message: 1:3
2018-03-26 16:49:25.841025 datelah[2324:754893] LOG SocketEngine: Got message: 3
2018-03-26 16:49:25.841827 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: pong with data: []
2018-03-26 16:49:25.842236 datelah[2324:754893] LOG SocketEnginePolling: Doing polling GET http://192.168.0.109:9005/socket.io/?transport=polling&b64=1&sid=cMDOxgYsL68hrm5AAAAZ
2018-03-26 16:49:25.888109 datelah[2324:754891] LOG SocketEngine: Got message: 3probe
2018-03-26 16:49:25.889765 datelah[2324:754891] LOG SocketEngine: Received probe response, should upgrade to WebSockets
2018-03-26 16:49:25.890016 datelah[2324:754891] LOG SocketEngine: Upgrading transport to WebSockets
2018-03-26 16:49:25.890223 datelah[2324:754891] LOG SocketEnginePolling: Sending poll:  as type: 6
2018-03-26 16:49:25.890702 datelah[2324:754891] LOG SocketEnginePolling: Created POST string: 1:6
2018-03-26 16:49:25.891583 datelah[2324:754891] LOG SocketEnginePolling: POSTing
2018-03-26 16:49:25.891949 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: pong with data: []
2018-03-26 16:49:25.986085 datelah[2324:754892] LOG SocketEnginePolling: Got polling response
2018-03-26 16:49:25.986449 datelah[2324:754892] LOG SocketEnginePolling: Got poll message: 1:6
2018-03-26 16:49:25.986791 datelah[2324:754892] LOG SocketEngine: Got message: 6
2018-03-26 16:49:25.987275 datelah[2324:754892] LOG SocketEngine: Switching to WebSockets
2018-03-26 16:49:25.987885 datelah[2324:754892] LOG SocketEngineWebSocket: Sending ws:  as type: 5
2018-03-26 16:49:25.988482 datelah[2324:754892] LOG SocketEngine: Flushing probe wait
2018-03-26 16:49:41.604850 datelah[2324:754465] righta arry is :(
    "type.."
)
2018-03-26 16:49:41.611460 datelah[2324:754465] LOG SocketIOClient{/}: Emitting: 2["new_message",{"userto":"1","message":"type..","userfrom":"2"}]
2018-03-26 16:49:41.612309 datelah[2324:754465] array is :(
        {
        message = "type..";
        userfrom = 2;
        userto = 1;
    }
)
2018-03-26 16:49:41.613690 datelah[2324:754926] LOG SocketEngine: Writing ws: 2["new_message",{"userto":"1","message":"type..","userfrom":"2"}] has data: false
2018-03-26 16:49:41.613994 datelah[2324:754926] LOG SocketEngineWebSocket: Sending ws: 2["new_message",{"userto":"1","message":"type..","userfrom":"2"}] as type: 4
2018-03-26 16:49:41.616259 datelah[2324:754465] rightcell
2018-03-26 16:49:41.625246 datelah[2324:754465] chat array value A:a
2018-03-26 16:49:48.454637 datelah[2324:754465] righta arry is :(
    "type..",
    Anitha
)
2018-03-26 16:49:48.455546 datelah[2324:754465] LOG SocketIOClient{/}: Emitting: 2["new_message",{"userto":"1","message":"Anitha","userfrom":"2"}]
2018-03-26 16:49:48.455895 datelah[2324:754465] array is :(
        {
        message = Anitha;
        userfrom = 2;
        userto = 1;
    }
)
2018-03-26 16:49:48.456571 datelah[2324:754937] LOG SocketEngine: Writing ws: 2["new_message",{"userto":"1","message":"Anitha","userfrom":"2"}] has data: false
2018-03-26 16:49:48.456810 datelah[2324:754937] LOG SocketEngineWebSocket: Sending ws: 2["new_message",{"userto":"1","message":"Anitha","userfrom":"2"}] as type: 4
2018-03-26 16:49:48.458609 datelah[2324:754465] rightcell
2018-03-26 16:49:48.459165 datelah[2324:754465] chat array value A:a
2018-03-26 16:49:48.470669 datelah[2324:754465] rightcell
2018-03-26 16:49:48.478191 datelah[2324:754465] chat array value A:a
2018-03-26 16:49:52.228034 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: ping with data: []
2018-03-26 16:49:52.228468 datelah[2324:754926] LOG SocketEngine: Writing ws:  has data: false
2018-03-26 16:49:52.230496 datelah[2324:754926] LOG SocketEngineWebSocket: Sending ws:  as type: 2
2018-03-26 16:49:52.472385 datelah[2324:754926] LOG SocketEngine: Got message: 3
2018-03-26 16:49:52.472850 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: pong with data: []
2018-03-26 16:49:52.885530 datelah[2324:754465] righta arry is :(
    "type..",
    Anitha,
    Hello
)
2018-03-26 16:49:52.886843 datelah[2324:754465] LOG SocketIOClient{/}: Emitting: 2["new_message",{"userto":"1","message":"Hello","userfrom":"2"}]
2018-03-26 16:49:52.887422 datelah[2324:754465] array is :(
        {
        message = Hello;
        userfrom = 2;
        userto = 1;
    }
)
2018-03-26 16:49:52.887547 datelah[2324:754926] LOG SocketEngine: Writing ws: 2["new_message",{"userto":"1","message":"Hello","userfrom":"2"}] has data: false
2018-03-26 16:49:52.887784 datelah[2324:754926] LOG SocketEngineWebSocket: Sending ws: 2["new_message",{"userto":"1","message":"Hello","userfrom":"2"}] as type: 4
2018-03-26 16:49:52.919632 datelah[2324:754465] rightcell
2018-03-26 16:49:52.919963 datelah[2324:754465] chat array value A:a
2018-03-26 16:49:52.922719 datelah[2324:754465] rightcell
2018-03-26 16:49:52.930987 datelah[2324:754465] chat array value A:a
2018-03-26 16:49:52.942004 datelah[2324:754465] rightcell
2018-03-26 16:49:52.949870 datelah[2324:754465] chat array value A:a
2018-03-26 16:50:19.716488 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: ping with data: []
2018-03-26 16:50:19.717041 datelah[2324:755029] LOG SocketEngine: Writing ws:  has data: false
2018-03-26 16:50:19.717564 datelah[2324:755029] LOG SocketEngineWebSocket: Sending ws:  as type: 2
2018-03-26 16:50:20.415392 datelah[2324:755029] LOG SocketEngine: Got message: 3
2018-03-26 16:50:20.416137 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: pong with data: []
2018-03-26 16:50:47.207898 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: ping with data: []
2018-03-26 16:50:47.208521 datelah[2324:755065] LOG SocketEngine: Writing ws:  has data: false
2018-03-26 16:50:47.209022 datelah[2324:755065] LOG SocketEngineWebSocket: Sending ws:  as type: 2
2018-03-26 16:50:47.282921 datelah[2324:754942] LOG SocketEngine: Got message: 3
2018-03-26 16:50:47.283670 datelah[2324:754465] LOG SocketIOClient{/}: Handling event: pong with data: []

**Comment by nuclearace** [comment]
It looks like the engine is never getting anything to begin with. Make sure that you're using your API correctly.

**Comment by Arunbalaji001** [comment]
This is code of the server, that I am running in my system. Sending and receiving via android client and web client works seamlessly. I am not receiving sockets in this iOS client only.

**Server**


       var app = require('express')();
       var http = require('http').Server(app);
        var socket = require('socket.io')(http);var port = 9005;




           socket.on('connection', function(client){
	       console.log('connected');
  
    
              client.on('join_user', function (roomid) {
	      var roomid = roomid;
	       client.join(roomid);
	       console.log('roomid: ' + roomid);
	      }); 

    
    
	client.on('new_message', function (data) {
	// we tell the client to execute 'new message'
	 var to_roomid = data.userto;
	 console.log('message:',data.message);
	//client.broadcast.emit('my_message', data); 
	client.to(to_roomid).emit('my_message', data); 
	console.log('toid:',to_roomid);

	});

      client.on('typing', function (data) {
	      var to_roomid = data.userto;

	       console.log("typing"+to_roomid);
         client.to(to_roomid).emit('typing',data);
       });

        // when the client emits 'stop typing', we broadcast it to others
          client.on('stop_typing', function (data) {
	  var to_roomid = data.userto;
	  console.log('stoptyping'+to_roomid);

             client.to(to_roomid).emit('stop_typing',data);
            });
    
             });


          http.listen(port, function(){
             console.log('listening on *:'+port);
           });

**Comment by Arunbalaji001** [comment]
Would be highly helpful to know if there is a way to listen to this server via iOS client.. does using rooms the server have something to do with this problem ?

**Comment by nuclearace** [comment]
Rooms are a server side feature only. Have you verified that messages are being received in the server?

**Comment by Arunbalaji001** [comment]
yes, messages are received by the server when sent from this client and forwarded to the correct rooms. After I sent the message to the server, I get the following in the Xcode log : 

```
2018-03-27 12:12:00.767136 datelah[2561:821843] LOG SocketIOClient{/}: Handling event: ping with data: []
2018-03-27 12:12:00.767652 datelah[2561:821843] Got event: ping, with items: (
)
2018-03-27 12:12:00.768062 datelah[2561:823602] LOG SocketEngine: Writing poll:  has data: false
2018-03-27 12:12:00.768482 datelah[2561:823602] LOG SocketEnginePolling: Sending poll:  as type: 2
```

when the client is at ideal I get the following in Xcode : 


```
2018-03-27 12:13:51.194206 datelah[2561:821843] LOG SocketManager: Engine opened Connect
2018-03-27 12:13:51.194403 datelah[2561:821843] LOG SocketIOClient{/}: Socket connected
2018-03-27 12:13:51.194610 datelah[2561:821843] LOG SocketIOClient{/}: Handling event: statusChange with data: [connected]
2018-03-27 12:13:51.194901 datelah[2561:821843] Got event: statusChange, with items: (
    connected
)
2018-03-27 12:13:51.195240 datelah[2561:821843] LOG SocketIOClient{/}: Handling event: connect with data: ["/"]
2018-03-27 12:13:51.195455 datelah[2561:821843] Got event: connect, with items: (
    "/"
)
2018-03-27 12:13:51.197243 datelah[2561:823723] LOG SocketEnginePolling: Sending poll:  as type: 2
2018-03-27 12:13:51.197572 datelah[2561:823723] LOG SocketEnginePolling: Created POST string: 1:21:21:2
2018-03-27 12:13:51.198433 datelah[2561:823723] LOG SocketEnginePolling: POSTing
2018-03-27 12:14:18.687768 datelah[2561:821843] LOG SocketIOClient{/}: Handling event: ping with data: []
2018-03-27 12:14:18.687971 datelah[2561:823839] LOG SocketEngine: Writing poll:  has data: false
2018-03-27 12:14:18.688314 datelah[2561:821843] Got event: ping, with items: (
)
2018-03-27 12:14:18.688486 datelah[2561:823839] LOG SocketEnginePolling: Sending poll:  as type: 2

```

**Comment by nuclearace** [comment]
@Arunbalaji001 Instead of calling `[manager emitAll:@"join_user" withItems:[SharedData sharedConstants].commonuserid];`, if you just call emit on the socket, does that work?

**Comment by Arunbalaji001** [comment]
Thank you.. After a bit of testing I found that the client is not emitting "join_user" event to server, due to which it was not receiving any messages sent to its room as the client have not joined into its specified room. Then I found that I have to use "[socket on  "connect",----------] " listener instead of "[socket on  "connection",--------] " to listen if the socket is connected. After I rectified that, the "join_user" event emitted fine and the client received the messages sent to its room perfectly.
Thank you for your time..
and I just want to say "socket.io" is a amazing library and the clients are working beautifully, it is really helpful to a budding developer like me.
